### ISSUE TITLE
Enhancement request: Access to "--diagnostics" equivalent via the API

### ISSUE THREAD

**Issue by mmorearty** [microsoft/TypeScript#19725]
**TypeScript Version:**  2.7.0-dev.20171103

Enhancement request: When running `tsc` from the command line, you can use `--diagnostics` to print some stats about the compilation — e.g. how many files were compiled; how many lines; how much time was taken by different phases of the compilation; and so on.

It would be nice if the same capabilities were exposed via the API. In other words, if I am writing a program that looks like this:

```ts
import * as ts from "typescript";

const program = ts.createProgram(...);
program.emit();
```

It would be nice if I could then write something like this on the next line:

```ts
const stats = ts.statistics();
```

Then I could do what I wanted with the stats — display them, analyze them, etc.

Currently, the `--diagnostics` functionality is entirely encapsulated inside `compiler/tsc.ts`, and as far as I can tell, that information is not available via the API.

**Comment by mhegazy** [comment]
The API allows for picking and choosing what to call and when. for instance you can call getSemanticDiagnostics on one file, but not all.. so in that sense getting the same output the compiler produces with `--diagnostics` does not make much sense..

the data collected is available, though not on the public API. i would be open to making some of these publicly available. eg. `Program.getNodeCount` and `program.getTypeCount` most of the numbers can be obtained from these.. see https://github.com/Microsoft/TypeScript/blob/78421e87f7a7d3c4f00bd54f3f01faaf4f6c30d7/src/compiler/tsc.ts#L197-L233

That said, I am interested to know what you need this information for?

**Comment by mmorearty** [comment]
tl;dr to get timings for a tsc wrapper that remains in memory and does incremental compiles, similar to `--watch` but a bit different. This feature request is just a nice-to-have, not a high priority.

The details: I've written a tool (to be open-sourced very soon) which is a "persistent worker" for the [Bazel](https://bazel.build/) build tool. With Bazel, a compiler or other build-related tool can be built to recognize a special command-line argument `--persistent_worker`, and then the tool can remain resident in memory, accepting build requests via stdin.

(`--watch` is obviously intended for the developer to use as he's coding, and is hard to use in a build tool such as Bazel or make or any other build tool. In our case we have a large number of Bazel "packages", and e.g. Package A might compile 20 .ts files and output `package_a.js` and `package_a.d.ts`; then Package B might depend on Package A, so it reads `package_a.d.ts`. Clearly `--watch` isn't the right tool for that.)

We've been using my tool for about six months here at Asana, with great success. Avoids having to repeatedly re-parse `.d.ts` files and so on.

Of course this saves on re-parsing of `lib.d.ts`, but in our case we also save on re-parsing of some large `.d.ts` files from our own codebase. (Some of the biggest ones are machine-generated from our database schema.) Plus we save on launch time of the compiler. (I'm giving a short talk about this this coming Monday, alongside a Google engineer who has some Bazel+TypeScript work of his own to share — see [here](https://sites.google.com/bazel.build/conference2017/schedule))

Anyway. That was long-winded :) It would be convenient to have my tool report how many lines of code were parsed and so on, in order to compare that to what `tsc` does.

**Comment by mmorearty** [comment]
Oh and you are certainly right to point out that when called via the API, you can pick and choose what gets done. E.g. in my sample code above, I showed a call to `ts.createProgram()`, but actually my code never does that; instead it calls `languageService.getProgram()`, thus getting better incrementality for my use case.

**Comment by mhegazy** [comment]
//cc @RyanCavanaugh, this set up would be interesting to project-to-project reference feature he is looking in now.

**Comment by mhegazy** [comment]
> tl;dr to get timings for a tsc wrapper that remains in memory and does incremental compiles, 

so why is not something as simple as `start = new Date(); program.getSemanticDiagnostics(); end - start`? the other info, like number of symbols, types, identifiers, etc.. do not seem to be of much value in this scenario really.

**Comment by mmorearty** [comment]
I've already done some simple timings similar to what you suggest. And I agree, symbols, types, identifiers, etc. are not of much value. I didn't know about `program.getSemanticDiagnostics()`, thanks, I'll look at that. 

I think the main thing I was interested in are:

* number of lines (but I realize there is ambiguity here — really I probably care about "number of lines _parsed_, which is different)
* Timings for I/O read/write and parsing

**Comment by mmorearty** [comment]
Ah, I see, the `program.getSemanticDiagnostics()` part of your suggestion was just, in a very hand-wavy way, the "compile" step, I get it

**Comment by mhegazy** [comment]
> number of lines (but I realize there is ambiguity here — really I probably care about "number of lines parsed, which is different)

You should know what files you are parsing on an update, we do have a lineMap property on the file, that should tell you how many lines are there. it is internal though, but we can expose it.

**Comment by mhegazy** [comment]
> Ah, I see, the program.getSemanticDiagnostics() part of your suggestion was just, in a very hand-wavy way, the "compile" step, I get it

not really.. the compile step is just parse/bind/check/emit.. 
`getSemanticDiagnostics()` is check.. 

here is what the commandline compiler does:

https://github.com/Microsoft/TypeScript/blob/78421e87f7a7d3c4f00bd54f3f01faaf4f6c30d7/src/compiler/tsc.ts#L165-L186
